plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUi.java

```diff
@@ -67,22 +92,173 @@ public void run() {
+      if (authorizationCode != null) {
+        return new VerificationCodeHolder(authorizationCode, redirectUrl);
+      }
+      return null;
+
+    } catch (IOException ioe) {
+      // Don't show an error dialog if a user pressed the cancel button.
+      if (!ioe.getMessage().contains(ERROR_MARKER_USER_CANCELLED_LOGIN)) {
+        showErrorDialogHelper(Messages.LOGIN_ERROR_DIALOG_TITLE,
+            Messages.LOGIN_ERROR_LOCAL_SERVER_RUN + ioe.getLocalizedMessage());
+      }
+      return null;
+    }
+  }
+
+  private String showProgressDialogAndWaitForCode(
+      final LocalServerReceiver codeReceiver, final String redirectUrl) throws IOException {
+    try {
+      final String[] codeHolder = new String[1];
+      final IOException[] exceptionHolder = new IOException[1];
+      final Semaphore wait = new Semaphore(0 /* initially zero permit */);
+
+      new ProgressMonitorDialog(shellProvider.getShell()) {
+        @Override
+        protected void configureShell(Shell shell) {
+          super.configureShell(shell);
+          shell.setText(Messages.LOGIN_PROGRESS_DIALOG_TITLE);
+        }
+        @Override
+        protected void cancelPressed() {
+          stopCodeWaitingJob(redirectUrl);
+          wait.release();  // Allow termination of the attached task.
+        }
+      }.run(true /* fork */, true /* cancelable */, new IRunnableWithProgress() {
+        @Override
+        public void run(IProgressMonitor monitor)
+            throws InvocationTargetException, InterruptedException {
+          monitor.beginTask(Messages.LOGIN_PROGRESS_DIALOG_MESSAGE, IProgressMonitor.UNKNOWN);
+          // Fork another sub-job to circumvent the limitation of LocalServerReceiver.
+          // (See the comments of scheduleCodeWaitingJob().)
+          scheduleCodeWaitingJob(
+              new LocalServerReceiverWrapper(codeReceiver), wait, codeHolder, exceptionHolder);
+          wait.acquire();  // Block until signaled.
+        }
+      });
+
+      if (exceptionHolder[0] != null) {
+        throw exceptionHolder[0];
+      }
+      return codeHolder[0];
+
+    } catch (InvocationTargetException | InterruptedException ex) {
+      // Never thrown from the attached task.
+      return null;
+    }
+  }
+
+  /**
+   * Schedule and run a job that calls {@link LocalServerReciever#waitForCode}. The reason for
+   * creating another job inside the job of {@link showProgressDialogAndWaitForCode} is that we
+   * cannot have a 100%-guarantee that {@link LocalServerReciever#waitForCode} will eventually
+   * return. (If {@link stopCodeWaitingJob} fails to stop {@link LocalServerReciever#waitForCode}
+   * gracefully, users will be stuck and the IDE has to be killed forcibly.)
+   *
+   * However, under normal circumstances, {@link stopCodeWaitingJob} will succeed to terminate
+   * this sub-job.
+   *
+   * @return scheduled job. Returning only for unit testing
```

Comment:
delete ". Returning only for unit testing" because no need to duplicate @VisibleForTesting